docs/proposals/14-wireguard-based-cluster-vpn.md

```diff
@@ -0,0 +1,178 @@
+
+As a mitigation we propose to instantiate 2 Kubelink instances in the seed cluster. The instances must run on different nodes (if possible but we assume a proper seed cluster has more than one node). Each shoot cluster connects to both endpoints. This means that the OpenVPN server is reachable with two different IP addresses. The VPN seed sidecars must attempt to connect to both of them and will continue to do so. The "Persistent Keepalive" feature is set to 21 seconds by default but could be reduced. Due to the redundancy this however appears not to be necessary.
+
+It is desirable that both connections are used in an equal manner. One strategy could be to use the kubelink 1 connection if the first target WireGuard address is even (the last byte of the IPv4 address), otherwise the kubelink 2 connection. The `vpn-seed` sidecars can then use the following configuration in their OpenVPN configuration file:
+
+ ```
+<connection>
+remote 192.168.45.3 1194 udp
+</connection>
+ 
+<connection>
+remote 192.168.47.34 1194 udp
+</connection>
+```
+
+OpenVPN will go through the list sequentially and try to connect to these endpoints.
+
+## Load Considerations
+
+What happens in case of a failure.? In this case one router will end up owning all connections as the clients will attempt to use the next connection. This could be mitigated by adding a third redundant WireGuard connection. Using this strategy, the failure of one WireGuard endpoint would result in the equal distribution of connections to the two remaining interfaces. We believe however that this will not be necessary.
+
+The cluster node running the Wireguard endpoint is essentially a router that routes all traffic to the various shoot clusters. This is established and proven technology that already exists since decades and has been highly optimized since then. This is also the technology that hyperscalers rely on to provide VPN connectivity to their customers. This said, hyperscalers essentially provide solutions based on IPsec which is known not to scale as well as Wireguard. Wireguard is a relatively new technology but we have no doubt that it is less stable than existing IPsec solution.
+
+We have some indications but no hard evidence that Wireguard performance and scalability does degrade with more than 1000 peers. This would mean that one single seed cluster should be able to serve roughly 1000 shoot clusters before running into issues. 
+
+Regarding performance there is a lot of information on the Internet basically suggesting that Wireguard performs better than other VPN solutions such as IPsec or OpenVPN. One example is https://www.wireguard.com/performance/.
+
+Based on this, we have no reason to believe that one router will not be able to handle all traffic going to and coming from shoot clusters. Nevertheless, we will closely monitor the situation in our tests and will take action if necessary.
```

Comment:
See above, it's hard to buy in if the argument is "we have no reason to believe".

Reply:
Well, this is how the Internet infrastructure works. Not sure how to phrase it in a different way...

Reply:
Does it make sense to evaluate this aspect beforehand and perform some scalability tests to get a better understanding of this assumption is valid or might break in the long run?

Reply:
> we will closely monitor the situation in our tests and will take action if necessary.

This should be fully automated. 

PS. I would not expect the wireguard website to have statement that any other competitive product is doing better.

Reply:
The performance should be bound to the resource (CPU, Memory, Network I/O) available to the underlying machine, isn't it? I guess it will perform differently with 1CPU/1GiB machine and 16CPU/64GiB machine.

Reply:
@vpnachev You are absolutely right. It depends on the system. The solution will introduce a slight performance hit due to the additional encryption performed by wireguard. It might be reduced if we disable the existing encryption for the vpn tunnel, which is currently performed in user space. The overhead in terms of Network I/O volume should be rather small as the existing traffic will simply be tunnelled. Additionally, the we enabled persistent connections in wireguard, which will add a udp packet every 21 second per shoot. Furthermore, the maximum transmission unit (MTU) will be slightly reduced.
In terms of memory, my understanding is that wireguard has a rather limited impact. That being said, we did not perform load tests, yet.

Reply:
closing here as we will go with a different approach.